### PR TITLE
Added CacheEntry in MorphShapeSymbol

### PR DESCRIPTION
**Pull request by gogoprog**
This change is [Reviewable](https://reviewable.io/reviews/fishingcactus/openfl/345)

### CODE REVIEW

**Comment by crazyjul** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[format/swf/lite/MorphShape.hx, line 123 at r1](https://reviewable.io:443/reviews/fishingcactus/openfl/345#-KvYqWPDTvEKauxNIL9b:-KvYqWPDTvEKauxNIL9c:b-yxn6qd) ([raw file](https://github.com/fishingcactus/openfl/blob/e641fa841aac37c5be5cdd8580b6c7970a8f26d8/format/swf/lite/MorphShape.hx#L123)):*
> ```Haxe
> 
>             } else {
>                 Reflect.setField(graphics, "__bitmap", cacheEntry.bitmapData);
> ```

As you target only js, you can go faster with (graphics:Dynamic).__bitmap.

This would not work

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fishingcactus/openfl/345)*

**Comment by gogoprog** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[format/swf/lite/MorphShape.hx, line 123 at r1](https://reviewable.io:443/reviews/fishingcactus/openfl/345#-KvYqWPDTvEKauxNIL9b:-KvYrFkiv6cOmy2hE0yQ:bh5t8zo) ([raw file](https://github.com/fishingcactus/openfl/blob/e641fa841aac37c5be5cdd8580b6c7970a8f26d8/format/swf/lite/MorphShape.hx#L123)):*
<details><summary><i>Previously, crazyjul (Julien Hamaide) wrote…</i></summary><blockquote>

As you target only js, you can go faster with (graphics:Dynamic).__bitmap.

This would not work
</blockquote></details>

What do you mean by 'not work' ? The current generated is graphics['__bitmap'] = ...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fishingcactus/openfl/345)*

**Comment by crazyjul** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[format/swf/lite/MorphShape.hx, line 123 at r1](https://reviewable.io:443/reviews/fishingcactus/openfl/345#-KvYqWPDTvEKauxNIL9b:-KvYrY_6cFYqybeZDTyV:b-3yt7ap) ([raw file](https://github.com/fishingcactus/openfl/blob/e641fa841aac37c5be5cdd8580b6c7970a8f26d8/format/swf/lite/MorphShape.hx#L123)):*
<details><summary><i>Previously, gogoprog (Gauthier Billot) wrote…</i></summary><blockquote>

What do you mean by 'not work' ? The current generated is graphics['__bitmap'] = ...
</blockquote></details>

My sentences was cut, "would not work on other platform"

If generated code is as said, nothing needs to be changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fishingcactus/openfl/345)*

**Comment by NuclearCookie** [comment]
:lgtm:

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/fishingcactus/openfl/345#-:-KvaeY_5EuwDpRDynvWk:bnfp4nl)*

File: format/swf/lite/MorphShape.hx [hunk 1]
```diff
@@ -84,31 +91,54 @@ class MorphShape extends Shape {
...
 	public override function __update (transformOnly:Bool, updateChildren:Bool):Void {
 
-		if(__renderDirty){
+        __updateTransforms();
 
-			var twips_ratio = Math.floor(ratio*65535);
-			var handler = __symbol.cachedHandlers.get(twips_ratio);
-			if ( handler == null ) {
-				var swf_shape = __symbol.getShape(ratio);
+		if(__renderDirty){
```
Comment (rams3s) [inline format/swf/lite/MorphShape.hx:96]: tabs ;p

File: format/swf/lite/MorphShape.hx [hunk 2]
```diff
@@ -84,31 +91,54 @@ class MorphShape extends Shape {
...
+                Reflect.setField(graphics, "__dirty", false);
+                @:privateAccess graphics.__bounds = cacheEntry.bounds;
+            }
+	    }
 
-				handler = new ShapeCommandExporter ();
-				swf_shape.export (handler);
-				__symbol.cachedHandlers.set(twips_ratio, handler);
-			}
+		super.__update(transformOnly, updateChildren);
```
Comment (rams3s) [inline format/swf/lite/MorphShape.hx:129]: I know it was like this before your changes, but can't we call super.__update() before processing the morph shape? That would save the first __updateTransforms() call

Comment (rams3s) [reply format/swf/lite/MorphShape.hx:129]: + tabs

File: openfl/display/Graphics.hx [hunk 3]
```diff
@@ -73,6 +73,7 @@ import js.html.CanvasRenderingContext2D;
 	private var __context:CanvasRenderingContext2D;
 	#end
 
+    public var keepBitmapData = false;
```
Comment (rams3s) [inline openfl/display/Graphics.hx:76]: tabs

**Comment by gogoprog** [comment]
*[format/swf/lite/MorphShape.hx, line 129 at r1](https://reviewable.io:443/reviews/fishingcactus/openfl/345#-KvXowO76PafbDC17-J3-r1-129:-KvfWJATGLJsRXZ4RXIJ:b-4sb06u) ([raw file](https://github.com/fishingcactus/openfl/blob/e641fa841aac37c5be5cdd8580b6c7970a8f26d8/format/swf/lite/MorphShape.hx#L129)):*
<details><summary><i>Previously, rams3s wrote…</i></summary><blockquote>

+ tabs
</blockquote></details>

It brings issues when calling super.__update() at first. Probably because bounds are not set yet. I haven't looked that much

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/fishingcactus/openfl/345)*

**Pull request merged**